File: Tests/AIKitTests/CostEstimatorTests.swift
import Foundation
import Testing
@testable import AIKit

// MARK: - CostEstimator tests

@Suite("CostEstimator.estimateTokens")
struct CostEstimatorTokenTests {

    @Test("Empty string returns 0")
    func emptyString() {
        #expect(CostEstimator.estimateTokens(text: "") == 0)
    }

    @Test("ASCII text: token count ≈ charCount/4 within ±20%")
    func asciiApproximation() {
        // 400 ASCII chars → expect ~100 tokens (80–120 is ±20%)
        let text = String(repeating: "abcd", count: 100) // 400 chars
        let tokens = CostEstimator.estimateTokens(text: text)
        #expect(tokens >= 80 && tokens <= 120)
    }

    @Test("Cyrillic text: token count ≈ charCount/3 within ±20%")
    func cyrillicApproximation() {
        // 300 Cyrillic chars → expect ~100 tokens (80–120 is ±20%)
        let text = String(repeating: "абвг", count: 75) // 300 chars
        let tokens = CostEstimator.estimateTokens(text: text)
        #expect(tokens >= 80 && tokens <= 120)
    }

    @Test("Single character returns at least 1")
    func singleChar() {
        #expect(CostEstimator.estimateTokens(text: "x") >= 1)
    }
}

@Suite("CostEstimator.estimate")
struct CostEstimatorCostTests {

    @Test("Zero tokens yields zero cost")
    func zeroCost() {
        let cost = CostEstimator.estimate(
            inputTokens: 0,
            outputTokens: 0,
            pricing: CostEstimator.anthropic_claude_sonnet_4_6
        )
        #expect(cost == 0.0)
    }

    @Test("Anthropic sonnet 4.6: 1000 input + 1000 output = $0.018")
    func anthropicSonnetCost() {
        // input: 1000/1M * $3.00 = $0.003
        // output: 1000/1M * $15.00 = $0.015
        // total: $0.018
        let cost = CostEstimator.estimate(
            inputTokens: 1000,
            outputTokens: 1000,
            pricing: CostEstimator.anthropic_claude_sonnet_4_6
        )
        #expect(abs(cost - 0.018) < 0.0001)
    }

    @Test("OpenAI gpt-4o-mini: 1000 input + 1000 output = $0.00075")
    func openAIMiniCost() {
        // input: 1000/1M * $0.15 = $0.00015
        // output: 1000/1M * $0.60 = $0.00060
        // total: $0.00075
        let cost = CostEstimator.estimate(
            inputTokens: 1000,
            outputTokens: 1000,
            pricing: CostEstimator.openai_gpt_4o_mini
        )
        #expect(abs(cost - 0.00075) < 0.000001)
    }

    @Test("Only input tokens, zero output")
    func inputOnlyTokens() {
        let cost = CostEstimator.estimate(
            inputTokens: 1_000_000,
            outputTokens: 0,
            pricing: CostEstimator.anthropic_claude_sonnet_4_6
        )
        #expect(abs(cost - 3.0) < 0.0001)
    }

    @Test("Only output tokens, zero input")
    func outputOnlyTokens() {
        let cost = CostEstimator.estimate(
            inputTokens: 0,
            outputTokens: 1_000_000,
            pricing: CostEstimator.anthropic_claude_sonnet_4_6
        )
        #expect(abs(cost - 15.0) < 0.0001)
    }
}

// MARK: - Transcription pricing

@Suite("CostEstimator.estimateTranscription")
struct CostEstimatorTranscriptionTests {

    @Test("60 s of audio at $0.006/min costs exactly $0.006")
    func oneMinuteWhisper1() {
        let cost = CostEstimator.estimateTranscription(
            durationSec: 60,
            pricing: CostEstimator.openai_whisper_1
        )
        #expect(abs(cost - 0.006) < 0.000001)
    }

    @Test("5-minute meeting at gpt-4o-mini-transcribe = $0.015")
    func fiveMinutesGpt4oMini() {
        let cost = CostEstimator.estimateTranscription(
            durationSec: 300,
            pricing: CostEstimator.openai_gpt_4o_mini_transcribe
        )
        #expect(abs(cost - 0.015) < 0.000001)
    }

    @Test("Zero-duration audio costs $0")
    func zeroDurationFree() {
        let cost = CostEstimator.estimateTranscription(
            durationSec: 0,
            pricing: CostEstimator.openai_whisper_1
        )
        #expect(cost == 0)
    }

    @Test("Negative duration is clamped to $0 (defensive)")
    func negativeDurationFree() {
        let cost = CostEstimator.estimateTranscription(
            durationSec: -10,
            pricing: CostEstimator.openai_whisper_1
        )
        #expect(cost == 0)
    }

    @Test("Sub-second durations scale proportionally")
    func subSecondScales() {
        let cost = CostEstimator.estimateTranscription(
            durationSec: 30,
            pricing: CostEstimator.openai_whisper_1
        )
        #expect(abs(cost - 0.003) < 0.000001)
    }

    @Test("gpt-4o-mini-transcribe is cheaper than gpt-4o-transcribe at the same duration")
    func miniIsCheaper() {
        let dur: Double = 600  // 10 minutes
        let mini = CostEstimator.estimateTranscription(durationSec: dur, pricing: CostEstimator.openai_gpt_4o_mini_transcribe)
        let full = CostEstimator.estimateTranscription(durationSec: dur, pricing: CostEstimator.openai_gpt_4o_transcribe)
        #expect(mini < full)
    }

    @Test("Deepgram Nova-2 batch is cheaper than whisper-1 at the same duration")
    func deepgramCheaperThanWhisper1() {
        let dur: Double = 600
        let deepgram = CostEstimator.estimateTranscription(durationSec: dur, pricing: CostEstimator.deepgram_nova_2_batch)
        let whisper = CostEstimator.estimateTranscription(durationSec: dur, pricing: CostEstimator.openai_whisper_1)
        #expect(deepgram < whisper)
    }
}

File: windows/tests/KosmoNotes.Core.Tests/CostEstimatorTests.cs
using KosmoNotes.Core.Pricing;

namespace KosmoNotes.Core.Tests;

public class CostEstimatorTests
{
    [Fact]
    public void EstimateTokens_EmptyString_IsZero()
    {
        Assert.Equal(0, CostEstimator.EstimateTokens(string.Empty));
    }

    [Fact]
    public void EstimateTokens_LatinUsesFourCharsPerToken()
    {
        // 16 chars / 4 = 4 tokens (ceiling).
        Assert.Equal(4, CostEstimator.EstimateTokens("the quick brown."));
    }

    [Fact]
    public void EstimateTokens_LatinShortString_AtLeastOne()
    {
        // 1 char / 4 = 0.25 → ceiling 1.
        Assert.Equal(1, CostEstimator.EstimateTokens("a"));
    }

    [Fact]
    public void EstimateTokens_UkrainianUsesThreeCharsPerToken()
    {
        // "Привіт" — 6 Cyrillic codepoints, 100% non-Latin → ratio 3.
        // 6 / 3 = 2 tokens.
        Assert.Equal(2, CostEstimator.EstimateTokens("Привіт"));
    }

    [Fact]
    public void EstimateTokens_MixedAtThreshold_PicksRatioBasedOnFraction()
    {
        // 7 latin + 3 cyrillic = 10 codepoints. 3/10 = 0.3 → NOT > 0.3 → 4-char ratio.
        // ceil(10/4) = 3 tokens.
        string mixed = "abcdefg" + "Бог";
        Assert.Equal(3, CostEstimator.EstimateTokens(mixed));

        // 6 latin + 4 cyrillic = 10 codepoints. 4/10 = 0.4 > 0.3 → 3-char ratio.
        // ceil(10/3) = 4 tokens.
        string mixed2 = "abcdef" + "Богі";
        Assert.Equal(4, CostEstimator.EstimateTokens(mixed2));
    }

    [Fact]
    public void Estimate_KnownTokenCounts_MatchesExpected()
    {
        // Anthropic Sonnet: $3 in / $15 out per 1M.
        // 1000 in + 500 out = 0.001 * 3 + 0.0005 * 15 = 0.003 + 0.0075 = 0.0105
        double cost = CostEstimator.Estimate(1000, 500, CostEstimator.AnthropicClaudeSonnet46);
        Assert.Equal(0.0105, cost, precision: 6);
    }

    [Fact]
    public void Estimate_OpenAIMiniIsCheap()
    {
        // 0.15 / 0.60 per 1M.
        // 10_000 in + 1_000 out = 0.01 * 0.15 + 0.001 * 0.60 = 0.0015 + 0.0006 = 0.0021
        double cost = CostEstimator.Estimate(10_000, 1_000, CostEstimator.OpenAIGpt4oMini);
        Assert.Equal(0.0021, cost, precision: 6);
    }

    [Fact]
    public void Estimate_ZeroTokens_IsZero()
    {
        Assert.Equal(0.0, CostEstimator.Estimate(0, 0, CostEstimator.OpenRouterDefault));
    }

    [Fact]
    public void EstimateTranscription_KnownDuration_MatchesExpected()
    {
        // Whisper-1: $0.006 / minute. 600 sec = 10 minutes → $0.06
        double cost = CostEstimator.EstimateTranscription(600, CostEstimator.OpenAIWhisper1);
        Assert.Equal(0.06, cost, precision: 6);

        // Deepgram Nova-2: $0.0043 / minute. 60 sec = 1 minute → $0.0043
        cost = CostEstimator.EstimateTranscription(60, CostEstimator.DeepgramNova2Batch);
        Assert.Equal(0.0043, cost, precision: 6);
    }

    [Fact]
    public void EstimateTranscription_NonPositiveDuration_IsZero()
    {
        Assert.Equal(0.0, CostEstimator.EstimateTranscription(0, CostEstimator.OpenAIWhisper1));
        Assert.Equal(0.0, CostEstimator.EstimateTranscription(-30, CostEstimator.OpenAIWhisper1));
    }

    [Fact]
    public void PricingTable_HasExpectedValues()
    {
        Assert.Equal(3.0, CostEstimator.AnthropicClaudeSonnet46.InputPerMillion);
        Assert.Equal(15.0, CostEstimator.AnthropicClaudeSonnet46.OutputPerMillion);

        Assert.Equal(0.15, CostEstimator.OpenAIGpt4oMini.InputPerMillion);
        Assert.Equal(0.60, CostEstimator.OpenAIGpt4oMini.OutputPerMillion);

        Assert.Equal(3.5, CostEstimator.OpenRouterDefault.InputPerMillion);
        Assert.Equal(17.0, CostEstimator.OpenRouterDefault.OutputPerMillion);

        Assert.Equal(0.006, CostEstimator.OpenAIWhisper1.PerMinute);
        Assert.Equal(0.006, CostEstimator.OpenAIGpt4oTranscribe.PerMinute);
        Assert.Equal(0.003, CostEstimator.OpenAIGpt4oMiniTranscribe.PerMinute);
        Assert.Equal(0.0043, CostEstimator.DeepgramNova2Batch.PerMinute);
    }
}
